### PR TITLE
fix: correct type definition for genReqId argument

### LINKED ISSUE #4748: Incorrect type definition for genReqId in FastifyServerOptions
### Prerequisites

- [X] I have written a descriptive issue title
- [X] I have searched existing issues to ensure the bug has not already been reported

### Fastify version

4.17.0

### Plugin version

_No response_

### Node.js version

Irrelevant

### Operating system

Linux

### Operating system version (i.e. 20.04, 11.3, 10)

Irrelevant

### Description

The type of `genReqId` server option [implies](https://github.com/fastify/fastify/blob/0dd31b71eeb79da4a43738473eaebcf5844fc8d4/fastify.d.ts#L112) its only argument is an instance of `FastifyRequest`:

```ts
    genReqId?: <...>(req: FastifyRequest<...>) => string,
```

However, in practice [it is given](https://github.com/fastify/fastify/blob/0dd31b71eeb79da4a43738473eaebcf5844fc8d4/lib/route.js#L397) `RawRequest` (e.g. `http.IncomingMessage`):

```js
function routeHandler (req, res, params, context, query) {
    const id = genReqId(req)
``` 

while `FastifyRequest` is [created only later](https://github.com/fastify/fastify/blob/0dd31b71eeb79da4a43738473eaebcf5844fc8d4/lib/route.js#L466) in this function.

### Steps to Reproduce

To reproduce, let's extend the test in [`test/genReqId.js`](https://github.com/fastify/fastify/blob/0dd31b71eeb79da4a43738473eaebcf5844fc8d4/test/genReqId.test.js#L11):

```js
    genReqId: function (req) {
      t.hasProp(req, 'id')
      t.hasProp(req, 'raw')
      return 'a'
    }
```
This test will now fail, because these two props normally found in `FastifyRequest` instance do not actually exist. (Well, `id` value is supposed to be provided by `genReqId` in the first place)

### Expected Behavior

The type definition should be something like this:

```ts
genReqId?: <...>(req: RawRequest) => string,
```

### PR DESCRIPTION
**Pull request by sergburn**
Custom `genReqId()` receives raw request as argument, not FastifyRequest, which is simply not created yet at the point when this function is called.`

#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

**Comment by sergburn** [comment]
Regarding failed "test (16, macos-latest)" job. It also fails on my old MacBook Air unless I disable any 2 tests in `404s.test.js` or simply reduce amount of tap jobs, e.g to 2. Something related to amount of open files?

File: docs/Reference/Server.md [hunk 1]
```diff
@@ -527,8 +527,8 @@ Defines the label used for the request identifier when logging the request.
 ### `genReqId`
 <a id="factory-gen-request-id"></a>
 
-Function for generating the request-id. It will receive the incoming request as
-a parameter. This function is expected to be error-free.
+Function for generating the request-id. It will receive _raw_ incoming
```
Comment (Fdawgs) [inline docs/Reference/Server.md:530]:
```suggestion
Function for generating the request-id. It will receive the _raw_ incoming
```

**Review by Fdawgs — CHANGES_REQUESTED** [review]

**Review by Fdawgs — APPROVED** [review]
Docs are a yes from me.
Need someone else to check the TS though.

File: test/types/fastify.test-d.ts [hunk 2]
```diff
@@ -127,7 +127,7 @@ expectAssignable<FastifyInstance>(fastify({ serverFactory: () => http.createServ
 expectAssignable<FastifyInstance>(fastify({ caseSensitive: true }))
 expectAssignable<FastifyInstance>(fastify({ requestIdHeader: 'request-id' }))
 expectAssignable<FastifyInstance>(fastify({ requestIdHeader: false }))
-expectAssignable<FastifyInstance>(fastify({ genReqId: () => 'request-id' }))
+expectAssignable<FastifyInstance>(fastify({ genReqId: (req) => req.rawHeaders[1] || 'foo' }))
```
Comment (Uzlopak) [inline test/types/fastify.test-d.ts:130]: The actual typing test needs to be something like this:

```typescript
fastify({ genReqId: (req) => {
  expectType<RawRequestDefaultExpression>(req)
}})
```

File: test/genReqId.test.js [hunk 3]
```diff
@@ -30,3 +30,43 @@ test('Should accept a custom genReqId function', t => {
...
+
+test('Custom genReqId function gets raw request as argument', t => {
+  t.plan(8)
+
+  const REQUEST_ID = 'REQ-1234'
+
+  const fastify = Fastify({
+    genReqId: function (req) {
+      t.notOk(req.id)
+      t.notOk(req.raw)
```
[the remark this answers is not in the document]
Comment (sergburn) [reply test/genReqId.test.js:42]: Maybe, this way is more specific?

```diff
   const fastify = Fastify({
     genReqId: function (req) {
-      t.notOk(req.id)
-      t.notOk(req.raw)
+      t.notOk('id' in req)
+      t.notOk('raw' in req)
+      t.ok(req instanceof Readable)
```

File: test/types/fastify.test-d.ts (hunk 2 above)
Comment (Uzlopak) [reply test/types/fastify.test-d.ts:130]: yeah I think they are unrelated to this change.

**Review by Uzlopak — COMMENTED** [review]
Sry, thought I answered your questions aleady.

File: test/types/fastify.test-d.ts (hunk 2 above)
Comment (sergburn) [reply test/types/fastify.test-d.ts]: Ok, I'll remove them

**Review by mcollina — APPROVED** [review]
lgtm

**Review by Uzlopak — APPROVED** [review]
LGTM

**Pull request merged**